staging/src/k8s.io/kube-apiextensions-server/test/integration/finalization_test.go

```diff
@@ -0,0 +1,89 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package integration
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	apiextensionsv1alpha1 "k8s.io/kube-apiextensions-server/pkg/apis/apiextensions/v1alpha1"
+	"k8s.io/kube-apiextensions-server/test/integration/testserver"
+)
+
+func TestFinalization(t *testing.T) {
+	stopCh, apiExtensionClient, clientPool, err := testserver.StartDefaultServer()
+	require.NoError(t, err)
+	defer close(stopCh)
+
+	noxuDefinition := testserver.NewNoxuCustomResourceDefinition(apiextensionsv1alpha1.ClusterScoped)
+	noxuVersionClient, err := testserver.CreateNewCustomResourceDefinition(noxuDefinition, apiExtensionClient, clientPool)
+	require.NoError(t, err)
+
+	ns := "not-the-default"
+	name := "foo123"
+	noxuResourceClient := NewNamespacedCustomResourceClient(ns, noxuVersionClient, noxuDefinition)
+
+	instance := testserver.NewNoxuInstance(ns, name)
+	instance.SetFinalizers([]string{"noxu.example.com/finalizer"})
+	createdNoxuInstance, err := instantiateCustomResource(t, instance, noxuResourceClient, noxuDefinition)
+	require.NoError(t, err)
+
+	uid := createdNoxuInstance.GetUID()
+	err = noxuResourceClient.Delete(name, &metav1.DeleteOptions{
+		Preconditions: &metav1.Preconditions{
+			UID: &uid,
+		},
+	})
+	require.NoError(t, err)
+
+	// Deleting something with a finalizer sets deletion timestamp to a not-nil value but does not
+	// remove the object from the API server. Here we read it to confirm this.
+	gottenNoxuInstance, err := noxuResourceClient.Get(name)
+	require.NoError(t, err)
+
+	require.NotNil(t, gottenNoxuInstance.GetDeletionTimestamp())
+
+	// Trying to delete it again to confirm it will not remove the object because finalizer is still there.
+	err = noxuResourceClient.Delete(name, &metav1.DeleteOptions{
+		Preconditions: &metav1.Preconditions{
+			UID: &uid,
+		},
+	})
+	require.NoError(t, err)
+
+	// Removing the finalizers to allow the following delete remove the object.
+	// This step will fail if previous delete wrongly removed the object.
+	gottenNoxuInstance.SetFinalizers(nil)
```

Comment:
That explains the spurious failures. I didn't know some other controllers will/may modify the object concurrently.
However then the question is why it was passing sometimes if line 64 always modifies it before the update? I tried to troubleshoot this earlier (see last block in comment https://github.com/kubernetes/kubernetes/pull/46439#issuecomment-304276161) and found that resource version was the same before and after the failed update operation so I thought there should be no conflict even though the conflict is reported. Weird. Delete does not increment the resource version but is checked on conditional update?
Anyway, I've updated the code. Thanks for the help!